From 52d5a3e7c51841d6f063412d86427c12ea007db2 Mon Sep 17 00:00:00 2001
From: Liang Li <liang.li@windriver.com>
Date: Tue, 10 Nov 2009 17:18:00 -0800
Subject: [PATCH] x86: Fix missed converted DMA address

Fault symptom:
--------------
Trying to use a device driver that uses SG (e.g. ATA) will result in
memory corruption and in the case of ATA, filesystem corruption.
Symptoms tend to correlate with random illogical Linux crashes e.g.

BUG: unable to handle kernel NULL pointer dereference at 00000010
IP: [<c01235be>] try_to_wake_up+0x2e/0x210
*pde = 00000000
Oops: 0000 [#1] PREEMPT SMP
LTT NESTING LEVEL : 0
Modules linked in:

Pid: 859, comm: sh Not tainted (2.6.27.37-WR3.0.2zz_standard-... #22)
EIP: 0060:[<c01235be>] EFLAGS: 00010002 CPU: 0
EIP is at try_to_wake_up+0x2e/0x210
EAX: 00000000 EBX: 00000000 ECX: 00000000 EDX: 00000000
ESI: ef3f0000 EDI: 00000000 EBP: 00000001 ESP: ef3f5d94
 DS: 007b ES: 007b FS: 00d8 GS: 0033 SS: 0068
 Process sh (pid: 859, ti=ef3f4000 task=ef8b66e0 task.ti=ef3f4000)
 Stack:
 00000001 00000000 00000000 ef36dc54 edf530e0 00000000 00000001 c011c3f3
 00000000 00000001 edf530ec edf530e8 00000000 00000001 00000292 c011dbde
 00000000 00000000 00000001 00000000 00000000 ef319807 edf53000 c02712c3
Call Trace:
[<c011c3f3>] __wake_up_common+0x43/0x70
[<c011dbde>] __wake_up+0x3e/0x60
[<c02712c3>] n_tty_receive_buf+0x5b3/0xfa0
[<c0241230>] blk_end_sync_rq+0x0/0x30
[<c017ee05>] kmem_cache_free+0x85/0xc0
--------------

Signed-off-by: Liang Li <liang.li@windriver.com>
---
 arch/x86/kernel/pci-nommu.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/pci-nommu.c b/arch/x86/kernel/pci-nommu.c
index ea0351c..12ea56a 100644
--- a/arch/x86/kernel/pci-nommu.c
+++ b/arch/x86/kernel/pci-nommu.c
@@ -75,7 +75,7 @@ static int nommu_map_sg(struct device *hwdev, struct scatterlist *sg,
 			dma_addr_t addr = sg_phys(s);
 
 			if (vbi_get_guest_dma_addr((void *)addr, &paddr) == 0)
-				s->dma_address = (dma_addr_t)paddr + s->offset;
+				s->dma_address = (dma_addr_t)paddr;
 			else
 				s->dma_address = -1;
 		} else
-- 
1.6.5.2

